File: apps/dsa-web/src/api/calculators.ts
// Copyright (c) 2026 SiinXu / StockPulse contributors
// SPDX-License-Identifier: AGPL-3.0-only
import { z } from 'zod';
import type { components } from '../types/api.generated';
import { parseCamelCasePayload } from './parseCamelCasePayload';
import apiClient from './index';

const BASE_PATH = '/api/v1/calculators';

type Schemas = components['schemas'];
type CamelCase<S extends string> = S extends `${infer Head}_${infer Tail}`
  ? `${Head}${Capitalize<CamelCase<Tail>>}`
  : S;
type CamelizeKeys<T> = T extends readonly (infer U)[]
  ? Array<CamelizeKeys<U>>
  : T extends object
    ? { [K in keyof T as CamelCase<K & string>]: CamelizeKeys<T[K]> }
    : T;

type OpenApiTargetContributionResponse =
  | Schemas['TargetContributionOkResponse']
  | Schemas['TargetContributionAlreadyMetResponse']
  | Schemas['TargetContributionUnreachableResponse'];
type OpenApiTargetDurationResponse =
  | Schemas['TargetDurationOkResponse']
  | Schemas['TargetDurationAlreadyMetResponse']
  | Schemas['TargetDurationUnreachableResponse'];

export type CompoundGrowthRequest = CamelizeKeys<Schemas['CompoundGrowthRequest']>;
export type TargetContributionRequest = CamelizeKeys<Schemas['TargetContributionRequest']>;
export type TargetDurationRequest = CamelizeKeys<Schemas['TargetDurationRequest']>;
export type CompoundGrowthResponse = CamelizeKeys<Schemas['CompoundGrowthResponse']>;
export type TargetContributionResponse = CamelizeKeys<OpenApiTargetContributionResponse>;
export type TargetDurationResponse = CamelizeKeys<OpenApiTargetDurationResponse>;

const balancePointSchema = z.object({
  period: z.number().int().nonnegative(),
  balance: z.number().finite(),
  totalContributed: z.number().finite(),
  gain: z.number().finite(),
}).strict();

const compoundGrowthResponseSchema: z.ZodType<CompoundGrowthResponse> = z.object({
  status: z.literal('ok'),
  principal: z.number().finite(),
  annualRate: z.number().finite(),
  years: z.number().finite(),
  contributionPerPeriod: z.number().finite(),
  periodsPerYear: z.number().int(),
  periodCount: z.number().int(),
  periodRate: z.number().finite(),
  finalValue: z.number().finite(),
  totalContributed: z.number().finite(),
  totalGain: z.number().finite(),
  seriesTotalPoints: z.number().int().min(2),
  seriesReturnedPoints: z.number().int().min(2).max(241),
  seriesSampled: z.boolean(),
  seriesStride: z.number().int().min(1),
  series: z.array(balancePointSchema).max(241),
}).strict();

const targetContributionBase = {
  target: z.number().finite(),
  principal: z.number().finite(),
  annualRate: z.number().finite(),
  years: z.number().finite(),
  periodsPerYear: z.number().int(),
  periodCount: z.number().int(),
  periodRate: z.number().finite(),
  currencyPrecisionDigits: z.literal(2),
  contributionRounding: z.literal('ceiling'),
};

const targetContributionResponseSchema: z.ZodType<TargetContributionResponse> = z.discriminatedUnion('status', [
  z.object({
    ...targetContributionBase,
    status: z.literal('ok'),
    reasonCode: z.literal('contribution_required'),
    contributionPerPeriod: z.number().finite(),
  }).strict(),
  z.object({
    ...targetContributionBase,
    status: z.literal('already_met'),
    reasonCode: z.literal('principal_growth_meets_target'),
    contributionPerPeriod: z.number().finite(),
  }).strict(),
  z.object({
    ...targetContributionBase,
    status: z.literal('unreachable'),
    reasonCode: z.literal('target_unreachable'),
    contributionPerPeriod: z.null(),
  }).strict(),
]);

const targetDurationBase = {
  target: z.number().finite(),
  principal: z.number().finite(),
  annualRate: z.number().finite(),
  contributionPerPeriod: z.number().finite(),
  periodsPerYear: z.number().int(),
  periodRate: z.number().finite(),
};

const targetDurationResponseSchema: z.ZodType<TargetDurationResponse> = z.discriminatedUnion('status', [
  z.object({
    ...targetDurationBase,
    status: z.literal('ok'),
    reasonCode: z.literal('duration_solved'),
    periodCount: z.number().int().positive(),
    years: z.number().positive().max(100),
  }).strict(),
  z.object({
    ...targetDurationBase,
    status: z.literal('already_met'),
    reasonCode: z.literal('principal_already_meets_target'),
    periodCount: z.literal(0),
    years: z.literal(0),
  }).strict(),
  z.object({
    ...targetDurationBase,
    status: z.literal('unreachable'),
    reasonCode: z.enum(['non_positive_trajectory', 'max_years_exceeded', 'target_unreachable']),
    periodCount: z.null(),
    years: z.null(),
  }).strict(),
]);

function toSnakeBody(body: Record<string, unknown>): Record<string, unknown> {
  return Object.fromEntries(Object.entries(body).map(([key, value]) => [
    key.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`),
    value,
  ]));
}

type RequestOptions = { signal?: AbortSignal };

export const calculatorsApi = {
  async compoundGrowth(
    body: CompoundGrowthRequest,
    options: RequestOptions = {},
  ): Promise<CompoundGrowthResponse> {
    const response = await apiClient.post(
      `${BASE_PATH}/compound-growth`,
      toSnakeBody(body as unknown as Record<string, unknown>),
      { signal: options.signal },
    );
    return parseCamelCasePayload<CompoundGrowthResponse>(
      response.data,
      compoundGrowthResponseSchema,
      'CompoundGrowthResponse',
    );
  },

  async targetContribution(
    body: TargetContributionRequest,
    options: RequestOptions = {},
  ): Promise<TargetContributionResponse> {
    const response = await apiClient.post(
      `${BASE_PATH}/target-contribution`,
      toSnakeBody(body as unknown as Record<string, unknown>),
      { signal: options.signal },
    );
    return parseCamelCasePayload<TargetContributionResponse>(
      response.data,
      targetContributionResponseSchema,
      'TargetContributionResponse',
    );
  },

  async targetDuration(
    body: TargetDurationRequest,
    options: RequestOptions = {},
  ): Promise<TargetDurationResponse> {
    const response = await apiClient.post(
      `${BASE_PATH}/target-duration`,
      toSnakeBody(body as unknown as Record<string, unknown>),
      { signal: options.signal },
    );
    return parseCamelCasePayload<TargetDurationResponse>(
      response.data,
      targetDurationResponseSchema,
      'TargetDurationResponse',
    );
  },
};

File: src/api/v1/schemas/calculators.py
# -*- coding: utf-8 -*-
"""Strict contracts for independent financial calculators (issue #240)."""

from __future__ import annotations

from typing import Annotated, List, Literal, Union

from pydantic import BaseModel, ConfigDict, Field


MAX_MONEY = 1e15
MAX_RATE = 10.0
MAX_YEARS = 100.0


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)


class CompoundGrowthRequest(_StrictModel):
    principal: float = Field(ge=0, le=MAX_MONEY, allow_inf_nan=False)
    annual_rate: float = Field(ge=-MAX_RATE, le=MAX_RATE, allow_inf_nan=False)
    years: float = Field(gt=0, le=MAX_YEARS, allow_inf_nan=False)
    contribution_per_period: float = Field(
        0.0,
        ge=-MAX_MONEY,
        le=MAX_MONEY,
        allow_inf_nan=False,
        description="End-of-period contribution; negative values are withdrawals",
    )
    periods_per_year: int = Field(12, ge=1, le=365)


class BalancePoint(_StrictModel):
    period: int = Field(ge=0)
    balance: float = Field(allow_inf_nan=False)
    total_contributed: float = Field(allow_inf_nan=False)
    gain: float = Field(allow_inf_nan=False)


class CompoundGrowthResponse(_StrictModel):
    status: Literal["ok"]
    principal: float = Field(allow_inf_nan=False)
    annual_rate: float = Field(allow_inf_nan=False)
    years: float = Field(allow_inf_nan=False)
    contribution_per_period: float = Field(allow_inf_nan=False)
    periods_per_year: int
    period_count: int
    period_rate: float = Field(allow_inf_nan=False)
    final_value: float = Field(allow_inf_nan=False)
    total_contributed: float = Field(allow_inf_nan=False)
    total_gain: float = Field(allow_inf_nan=False)
    series_total_points: int = Field(ge=2)
    series_returned_points: int = Field(ge=2, le=241)
    series_sampled: bool
    series_stride: int = Field(ge=1)
    series: List[BalancePoint]


class TargetContributionRequest(_StrictModel):
    target: float = Field(ge=0, le=MAX_MONEY, allow_inf_nan=False)
    principal: float = Field(ge=0, le=MAX_MONEY, allow_inf_nan=False)
    annual_rate: float = Field(ge=-MAX_RATE, le=MAX_RATE, allow_inf_nan=False)
    years: float = Field(gt=0, le=MAX_YEARS, allow_inf_nan=False)
    periods_per_year: int = Field(12, ge=1, le=365)


class _TargetContributionBase(_StrictModel):
    target: float = Field(allow_inf_nan=False)
    principal: float = Field(allow_inf_nan=False)
    annual_rate: float = Field(allow_inf_nan=False)
    years: float = Field(allow_inf_nan=False)
    periods_per_year: int
    period_count: int
    period_rate: float = Field(allow_inf_nan=False)
    currency_precision_digits: Literal[2]
    contribution_rounding: Literal["ceiling"]


class TargetContributionOkResponse(_TargetContributionBase):
    status: Literal["ok"]
    reason_code: Literal["contribution_required"]
    contribution_per_period: float = Field(allow_inf_nan=False)


class TargetContributionAlreadyMetResponse(_TargetContributionBase):
    status: Literal["already_met"]
    reason_code: Literal["principal_growth_meets_target"]
    contribution_per_period: float = Field(allow_inf_nan=False)


class TargetContributionUnreachableResponse(_TargetContributionBase):
    status: Literal["unreachable"]
    reason_code: Literal["target_unreachable"]
    contribution_per_period: None = None


TargetContributionResponse = Annotated[
    Union[
        TargetContributionOkResponse,
        TargetContributionAlreadyMetResponse,
        TargetContributionUnreachableResponse,
    ],
    Field(discriminator="status"),
]


class TargetDurationRequest(_StrictModel):
    target: float = Field(ge=0, le=MAX_MONEY, allow_inf_nan=False)
    principal: float = Field(ge=0, le=MAX_MONEY, allow_inf_nan=False)
    annual_rate: float = Field(ge=-MAX_RATE, le=MAX_RATE, allow_inf_nan=False)
    contribution_per_period: float = Field(
        ge=-MAX_MONEY,
        le=MAX_MONEY,
        allow_inf_nan=False,
        description="End-of-period contribution; negative values are withdrawals",
    )
    periods_per_year: int = Field(12, ge=1, le=365)


class _TargetDurationBase(_StrictModel):
    target: float = Field(allow_inf_nan=False)
    principal: float = Field(allow_inf_nan=False)
    annual_rate: float = Field(allow_inf_nan=False)
    contribution_per_period: float = Field(allow_inf_nan=False)
    periods_per_year: int
    period_rate: float = Field(allow_inf_nan=False)


class TargetDurationOkResponse(_TargetDurationBase):
    status: Literal["ok"]
    reason_code: Literal["duration_solved"]
    period_count: int = Field(ge=1)
    years: float = Field(gt=0, le=MAX_YEARS, allow_inf_nan=False)


class TargetDurationAlreadyMetResponse(_TargetDurationBase):
    status: Literal["already_met"]
    reason_code: Literal["principal_already_meets_target"]
    period_count: Literal[0]
    years: float = Field(ge=0, le=0, allow_inf_nan=False)


class TargetDurationUnreachableResponse(_TargetDurationBase):
    status: Literal["unreachable"]
    reason_code: Literal[
        "non_positive_trajectory",
        "max_years_exceeded",
        "target_unreachable",
    ]
    period_count: None = None
    years: None = None


TargetDurationResponse = Annotated[
    Union[
        TargetDurationOkResponse,
        TargetDurationAlreadyMetResponse,
        TargetDurationUnreachableResponse,
    ],
    Field(discriminator="status"),
]
